### DISCUSSION TITLE
Use Artillery with existing nodejs test scripts

### DISCUSSION THREAD

**Discussion by tchambers97** [artilleryio/artillery#1428] [Artillery]
TL;DR: Is there a way to use artillery to call our existing nodejs API tests?

We have a rather extensive nodejs library of API tests that we run with the Jest framework. We are interested in integrating these with artillery to create load tests. I was hoping this could be done without having to redo all this work in artillery yaml files. My first thought was to use artillery as a nodejs library within our existing tests, but from other reading in this forum, I gather that's not recommended.

**Comment by GeekyDeaks** [comment]
I am doing something similar.  Essentially, using artillery as the end-to-end testing as well as load testing.  I found what I needed to launch artillery programmatically by reviewing the tests e.g. https://github.com/artilleryio/artillery/blob/master/test/core/test_capture.js

I just replaced the mocking with an instance of the service I wanted to test.  The current approach I am working on is to keep each scenario in it's own file so that a single instance of each can be ran for unit-testing, but then they can be combined into a load test flow later in the CI pipeline.

**Comment by hassy** [comment]
great answer @GeekyDeaks!

@tchambers97 another option is to call your existing scripts from a `function` step in an Artillery scenario:

https://www.artillery.io/docs/guides/guides/http-reference#function-steps-in-a-flow

It depends on how your current tests are structured, but you should also be able to make use of the Artillery's extension APIs (https://www.artillery.io/docs/guides/guides/extension-apis#tracking-custom-metrics) to track metrics from that code to get reports from Artillery on number of requests made, latency distribution etc.

The [Playwright engine](https://github.com/artilleryio/artillery-engine-playwright) may be of interest. It does something like that to wrap Playwright scripts and run them as load tests.

**Reply by tchambers97** [reply]
Thank you! The function step looks promising to someone with limited JS abilities like me.

**Comment by GeekyDeaks** [comment]
re-reading this again @tchambers97 , I think the idea suggested by @hassy about creating your own engine is a much better place to start looking than my suggestion.  I came across this article for running jest tests programmatically https://medium.com/web-developers-path/how-to-run-jest-programmatically-in-node-js-jest-javascript-api-492a8bc250de - the runCLI is still being exposed, so it might be worth checking out to see if you can wrap a simple artillery engine around it

**Reply by tchambers97** [reply]
Thank you! I will give both suggestions a look.
